Examples code updated with changes as requested (default values removed, use of variables restricted to only least, local values used wherever applicable, provider with lowest and one latest version defined, test/plan/apply verified)

marifse opened the pull request:
### Description

<!--- The PR is being raised to make the examples terraform code not directly consumeable, in this regard the default values for prefix and region have been removed, besides that the basic example is already using the lowest available provider and also have tested running the terraform plan and apply --->

### Release required?
<!--- Identify the type of release. For information about the changes in a semantic versioning release, see [Release versioning](https://terraform-ibm-modules.github.io/documentation/#/versioning). --->

- [ ] No release
- [ ] Patch release (`x.x.X`)
- [ ] Minor release (`x.X.x`)
- [ ] Major release (`X.x.x`)

##### Release notes content

<!--- If a release is required, replace this text with information that users need to know about the release. Write the release notes to help users understand the changes, and include information about how to update from the previous version.

Your notes help the merger write the commit message for the PR that is published in the release notes for the module. --->

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### For mergers

- Use a conventional commit message to set the release level. Follow the [guidelines](https://terraform-ibm-modules.github.io/documentation/#/merging.md).
- Include information that users need to know about the PR in the commit message. The commit message becomes part of the GitHub release notes.
- Use the **Squash and merge** option.

marifse commented:
/run pipeline

marifse commented:
@ocofaigh can you please check and approve if okay.

ocofaigh reviewed (changes requested):
There is only a basic example here which tests the lowest provider version. Can you create an "advanced" example which will test latest provider version. It should support the following:
- resource group creation
- key protect instance with
  - `service_endpoints` set to private
  - pass some values for `key_create_import_access_settings`
- create new key ring (using https://github.com/terraform-ibm-modules/terraform-ibm-kms-key-ring)
- create new key in new key ring (https://github.com/terraform-ibm-modules/terraform-ibm-kms-key)

You should add a new test for the new example in pr_test.go and you should update the cra-config.yaml to now scan the new example

marifse commented:
> st latest provider version. It should support the following:

sure. let me

marifse commented:
> There is only a basic example here which tests the lowest provider version. Can you create an "advanced" example which will test latest provider version. It should support the following:
> 
> * resource group creation
> * key protect instance with
>   
>   * `service_endpoints` set to private
>   * pass some values for `key_create_import_access_settings`
> * create new key ring (using https://github.com/terraform-ibm-modules/terraform-ibm-kms-key-ring)
> * create new key in new key ring (https://github.com/terraform-ibm-modules/terraform-ibm-kms-key)
> 
> You should add a new test for the new example in pr_test.go and you should update the cra-config.yaml to now scan the new example

@ocofaigh can you please check now, i have added the advance example, tested it, though i have tested it with public access as for private i have to be IBM cloud private network. So i added a comment in main.tf, you can review and let me know your thoughts. Also i have added the test in pre-test.go but i do not try running that.

marifse commented:
@ocofaigh i have a query, if i am removing some variable default value, then i am supposed to add the variable in catalogValidationValues.json.template.json, with what value "hard coded" or like this $variablevalue, like in another PGSQL example, i removed the default value for admin_pass variable, and i have to include it in catalogValidationValues.json.template.json, it below is okay? and if okay where do i need to pass the value so that the test can run successfull.

{
  "ibmcloud_api_key": $VALIDATION_APIKEY,
  "region": "us-south",
  "resource_tags": $TAGS,
  "prefix": $PREFIX,
  "admin_pass": $ADMIN_PASSWORD
}

Also where are the values for these $ are getting fetched when running the tests, or i need to pass them in test option.

ocofaigh commented on examples/advance/main.tf:
```diff
@@ -0,0 +1,52 @@
+  region            = var.region
+  tags              = var.resource_tags
+  access_tags       = var.access_tags
+  #service_endpoints  = "private-only"      #Uncomment this line to create the Private Key Protect, this will require you to be part of IBM cloud private network when creating this resource
```
set this to `service_endpoints = "private"` - we can write a test that actually tests this test in IBM Schematics which will have access to the private network (see subsequent comment)

ocofaigh commented on examples/advance/version.tf:
```diff
@@ -0,0 +1,10 @@
+terraform {
+  required_version = ">= 1.0.0, <1.6.0"
+  required_providers {
+    # Pin to the lowest provider version of the range defined in the main module's version.tf to ensure lowest version still works
```
This comment is not valid. Please update to:

```
# Ensure that there is always 1 example locked into the lowest provider version of the range defined in the main
# module's version.tf (basic example), and 1 example that will always use the latest provider version (advanced example).
```

ocofaigh commented on tests/pr_test.go:
```diff
@@ -54,6 +55,22 @@ func TestRunBasicExample(t *testing.T) {
 	assert.NotNil(t, output, "Expected some output")
 }
 
+func TestRunAdvanceExample(t *testing.T) {
```
So in order to test the advanced example, lets copy the approach [here](https://github.com/terraform-ibm-modules/terraform-ibm-base-ocp-vpc/blob/0295160a963d3beb60394bfd60da6a42513f83f1/tests/pr_test.go#L87-L127) which will deploy the example in schematics which means the private endpoint can be used.

ocofaigh commented on examples/advance/README.md:
```diff
@@ -0,0 +1,9 @@
+# Advance example
+
+A advance example that shows how to provision a Private Key Protect instance with Key Ring.
```
An advanced example that shows how to provision a private only Key Protect instance with a new Key Ring and Key.

ocofaigh commented on examples/advance/README.md:
```diff
@@ -0,0 +1,9 @@
+
+The following resources are provisioned by this example:
+ - A new resource group, if an existing one is not passed in.
+ - A new Private Key Protect instance in the given resource group and region.
```
Private -> private

ocofaigh commented on examples/advance/main.tf:
```diff
@@ -0,0 +1,52 @@
+}
+
+##############################################################################
+# Key Protect with Private Service Endpoint and Keyring
```
remove "and Keyring"

ocofaigh commented on examples/advance/main.tf:
```diff
@@ -0,0 +1,52 @@
+  source      = "terraform-ibm-modules/kms-key-ring/ibm"
+  version     = "2.3.1"
+  instance_id = module.key_protect_module.key_protect_guid
+  key_ring_id = "my-key-ring"
```
use the prefix for naming. EG: `"${var.prefix}-my-key-ring"`

ocofaigh commented on examples/advance/main.tf:
```diff
@@ -0,0 +1,52 @@
+  version     = "2.3.1"
+  instance_id = module.key_protect_module.key_protect_guid
+  key_ring_id = "my-key-ring"
+  depends_on  = [module.key_protect_module]
```
Don't use `depends_on` - there is already an implicit dependency with the instance GUID

ocofaigh reviewed (changes requested):
I left some comments. Could you also please add the new advanced exampled to the ibm_catalog.json in the flavors array here -> https://github.com/marifse/terraform-ibm-key-protect/blob/52f7bb3daafe237c9c5ceb81961680067d366b15/ibm_catalog.json#L54

In terms of the variables in the `catalogValidationValues.json.template.json` file, they are not used by the PR pipeline, they are used by an internal pipeline AFTER code is merged. I can assist you with any new required values as part of the PR review process. Unfortunately the full list of them are only documented internally since its an internal only pipeline.

marifse commented on examples/advance/main.tf:
```diff
@@ -0,0 +1,52 @@
+}
+
+##############################################################################
+# KMS standard key
```
@ocofaigh requested changes have been addresses, except the pre-test.go, i am not sure either that is correct or not. can you please look into that and let me know further. thanks

ocofaigh commented on tests/pr_test.go:
```diff
@@ -54,6 +57,43 @@ func TestRunBasicExample(t *testing.T) {
+	if options.Region == "jp-osa" {
+		options.Region = "us-south"
+	}
+
```
There is currently a provider bug which this test will hit when using private only instance. To workaround it, please add this:

```
// Workaround for https://github.com/IBM-Cloud/terraform-provider-ibm/issues/5154
options.AddWorkspaceEnvVar("IBMCLOUD_KP_API_ENDPOINT", "https://private."+region+".kms.cloud.ibm.com", false, false)
```

ocofaigh reviewed (changes requested):
see comments

marifse commented on examples/advance/README.md:
```diff
@@ -0,0 +1,9 @@
+ - A new resource group, if an existing one is not passed in.
+ - A new Private Key Protect instance in the given resource group and region.
+ - A new KMS Key Ring.
+ - A new KMS Key.
```
done

marifse commented:
> I left some comments. Could you also please add the new advanced exampled to the ibm_catalog.json in the flavors array here -> https://github.com/marifse/terraform-ibm-key-protect/blob/52f7bb3daafe237c9c5ceb81961680067d366b15/ibm_catalog.json#L54
> 
> In terms of the variables in the `catalogValidationValues.json.template.json` file, they are not used by the PR pipeline, they are used by an internal pipeline AFTER code is merged. I can assist you with any new required values as part of the PR review process. Unfortunately the full list of them are only documented internally since its an internal only pipeline.

@ocofaigh please check now.

ocofaigh commented on tests/pr_test.go:
```diff
@@ -54,6 +57,40 @@ func TestRunBasicExample(t *testing.T) {
+	// "jp-osa" is currently not allowing hs-crypto be used for encrypting in that region.
+	if options.Region == "jp-osa" {
+		options.Region = "us-south"
+	}
```
There is no need for this logic here as we are not using HPCS in this example

marifse replied:
@ocofaigh removed this logic

marifse replied:
@ocofaigh do have any clue on golang_lint, how can i fix it. It is also failing on my side even after updating the golang_lint

ocofaigh commented:
/run pipeline

ocofaigh reviewed (changes requested):
Hmm, the golang linter is failing in the pipeline. Are you able to reproduce it locally by running `pre-commit run --all-files ` ?

Here is what pipeline is failing with: `level=error msg="Running error: can't run linter goanalysis_metalinter\nbuildir: failed to load package : could not load export data: no export data for \"github.com/IBM/schematics-go-sdk/schematicsv1\""`

ocofaigh commented:
/run pipeline

ocofaigh commented on tests/pr_test.go:
```diff
@@ -6,14 +6,17 @@ import (
 	"os"
 	"testing"
 
+	"github.com/terraform-ibm-modules/ibmcloud-terratest-wrapper/testschematic"
```
@marifse maybe try moving this as the last import? Just a guess, but that works in [this test](https://github.com/terraform-ibm-modules/terraform-ibm-kms-all-inclusive/blob/bcdb164522ff654f674fa7bef717a97f5b1aa5f9/tests/pr_test.go#L4-L13)

ocofaigh commented:
@marifse It looks like you removed the test in your latest commit - we need a test for the advanced example

marifse commented:
@ocofaigh test added for advanced example again, and also the go_lint error fixed. Please review and re-run the pipeline.

marifse commented on tests/pr_test.go:
```diff
@@ -54,6 +55,35 @@ func TestRunBasicExample(t *testing.T) {
+		DeleteWorkspaceOnFail:  false,
+		WaitJobCompleteMinutes: 60,
+	})
+
```
@ocofaigh after adding this line, i started to get the below error:

pr_test.go:1: : # github.com/terraform-ibm-modules/terraform-ibm-key-protect [github.com/terraform-ibm-modules/terraform-ibm-key-protect.test]
./pr_test.go:77:76: undefined: region (typecheck)

Before this the CI ran successful

marifse replied:
@ocofaigh can you please try running the pipeline now.

ocofaigh replied:
@marifse thats because you don't have any region variable defined. Try this:
```
// Workaround for https://github.com/IBM-Cloud/terraform-provider-ibm/issues/5154 options.AddWorkspaceEnvVar("IBMCLOUD_KP_API_ENDPOINT", "https://private."+options.Region+".kms.cloud.ibm.com", false, false)
```

marifse replied:
@ocofaigh done. I also tried that but that time it did not work at first. Anyways now it is okay.

ocofaigh commented:
/run pipeline

ocofaigh commented on tests/pr_test.go:
```diff
@@ -54,6 +55,38 @@ func TestRunBasicExample(t *testing.T) {
+		Prefix:  "advanced-key-protect",
+		TarIncludePatterns: []string{
+			"*.tf",
+			"modules/*/*.tf",
```
there is no modules folder in this repo so I think thats why test failed in pipeline. You will want to add `advancedExampleTerraformDir` here

marifse replied:
@removed this folder. please re-run

ocofaigh commented on tests/pr_test.go:
```diff
@@ -54,6 +55,37 @@ func TestRunBasicExample(t *testing.T) {
+		Testing: t,
+		Prefix:  "advanced-key-protect",
+		TarIncludePatterns: []string{
+			"*.tf",
```
you need to also add the example directory (aka `advancedExampleTerraformDir`)

marifse replied:
@ocofaigh done

ocofaigh commented on tests/pr_test.go:
```diff
@@ -54,6 +55,37 @@ func TestRunBasicExample(t *testing.T) {
+		Testing: t,
+		Prefix:  "advanced-key-protect",
+		TarIncludePatterns: []string{
+			advancedExampleTerraformDir + "/*.tf",
```
please add back the root level - the example calls the root level module so both are needed in the tar file that gets sent to schematics
```
TarIncludePatterns: []string{
			"*.tf",
			advancedExampleTerraformDir + "/*.tf",
		},
```

marifse replied:
@ocofaigh done.

ocofaigh commented:
/run pipeline
